rust/lance/src/dataset/write/commit.rs

```diff
@@ -0,0 +1,339 @@
+                        location: location!(),
+                    });
+                }
+            }
+        }
+
+        let manifest_config = ManifestWriteConfig {
+            use_move_stable_row_ids,
+            storage_format: self.storage_format.map(DataStorageFormat::new),
+            ..Default::default()
+        };
+
+        let manifest = if let Some(dataset) = dest.dataset() {
+            if self.detached {
+                if matches!(manifest_naming_scheme, ManifestNamingScheme::V1) {
+                    return Err(Error::NotSupported {
+                        source: "detached commits cannot be used with v1 manifest paths".into(),
+                        location: location!(),
+                    });
+                }
+                commit_detached_transaction(
+                    dataset,
+                    object_store.as_ref(),
+                    commit_handler.as_ref(),
+                    &transaction,
+                    &manifest_config,
+                    &self.commit_config,
+                )
+                .await?
+            } else {
+                commit_transaction(
+                    dataset,
+                    object_store.as_ref(),
+                    commit_handler.as_ref(),
+                    &transaction,
+                    &manifest_config,
+                    &self.commit_config,
+                    manifest_naming_scheme,
+                )
+                .await?
+            }
+        } else if self.detached {
+            // I think we may eventually want this, and we can probably handle it, but leaving a TODO for now
+            return Err(Error::NotSupported {
+                source: "detached commits cannot currently be used to create new datasets".into(),
+                location: location!(),
+            });
+        } else {
+            commit_new_dataset(
+                object_store.as_ref(),
+                commit_handler.as_ref(),
+                &base_path,
+                &transaction,
+                &manifest_config,
+                manifest_naming_scheme,
+            )
+            .await?
+        };
+
+        let tags = Tags::new(
+            object_store.clone(),
+            commit_handler.clone(),
+            base_path.clone(),
+        );
+
+        match &self.dest {
+            WriteDestination::Dataset(dataset) => Ok(Dataset {
+                manifest: Arc::new(manifest),
+                session: self.session.unwrap_or(dataset.session.clone()),
+                ..dataset.as_ref().clone()
+            }),
+            WriteDestination::Uri(uri) => Ok(Dataset {
+                object_store,
+                base: base_path,
+                uri: uri.to_string(),
+                manifest: Arc::new(manifest),
+                session: self.session.unwrap_or_default(),
+                commit_handler,
+                tags,
+                manifest_naming_scheme,
+            }),
+        }
+    }
+}
```

Comment:
We have enough unit tests around this method that I'll assume this was ported over correctly :laughing:

Reply:
Yeah I did a lot of work to get it passing with the existing tests. I feel pretty confident in it.